File: Algorithms/0792_Number_of_Matching_Subsequences/Java/NumberofMatchingSubsequencesSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private Map<String, Integer> result = new HashMap<>();

    public int numMatchingSubseq(String s, String[] words) {
        int res = 0;
        for (String word : words) {
            if (result.containsKey(word)) res += result.get(word);
            else {
                if (isSubsequence(word, s)) {
                    res += 1;
                    result.put(word, 1);
                } else {
                    result.put(word, 0);
                }
            }
        }
        return res;
    }

    private boolean isSubsequence(String s, String target) {
        int index1 = 0, index2 = 0;
        while (index1 < s.length() && index2 < target.length()) {
            if (s.charAt(index1) == target.charAt(index2)) index1++;
            if (index1 == s.length()) return true;
            index2++;
        }
        return index1 == s.length();
    }
}

File: Algorithms/0792_Number_of_Matching_Subsequences/Python/Number_of_Matching_Subsequences_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def numMatchingSubseq(self, s, words):
        res = 0
        self.result = {}

        for word in words:
            if word in self.result:
                res += self.result[word]
            else:
                if self.is_subsequence(s, word):
                    res += 1
        return res

    def is_subsequence(self, target, s):
        length1 = len(target)
        length2 = len(s)
        index1, index2 = 0, 0
        while index1 < length1 and index2 < length2:
            if target[index1] == s[index2]: index2 += 1
            if index2 == length2:
                self.result[s] = 1
                return True
            index1 += 1
        self.result[s] = 0
        return False





